iwyu_output.cc

```diff
@@ -605,6 +619,29 @@ void IwyuFileInfo::ReportForwardDeclareUse(SourceLocation use_loc,
   LogSymbolUse("Marked fwd-decl use of decl", symbol_uses_.back());
 }
 
+void IwyuFileInfo::ReportUsingDeclUse(SourceLocation use_loc,
+                                      const UsingDecl* using_decl,
+                                      bool in_cxx_method_body,
+                                      const char* comment) {  
+
+  // If accessing a symbol through a using decl in the same file that contains
+  // the using decl, we must mark the using decl as referenced. At the end of
+  // traversing the AST, we check to see if a using decl is unreferenced and a
+  // full use of one of its shadow decls so that the source file continues to
+  // compile.
+  typename map<const UsingDecl*, bool>::iterator using_decl_status = 
+    using_decl_referenced_.find(using_decl);
+
+  if(using_decl_status != using_decl_referenced_.end()) {
+    using_decl_status->second = true;
+  }
+
+  // When a symbol is accessed through a using decl, we must report
+  // that as a full use of the using decl because whatever file that
+  // using decl is in is now required.
+  ReportFullSymbolUse(use_loc, using_decl, in_cxx_method_body, comment);
```

Comment:
What if we also add that using_decl is making ForwardDeclareUse of shadow target? I.e. something like `ReportForwardDeclareUse(using_decl->getUsingLoc(), decl, ...)`? I haven't found a case where it is crucial, because `decl` is marked as used anyway. But seems like it is correct thing to do.